vsintegration/tests/FSharp.Editor.Tests/CodeFixes/CodeFixTestFramework.fs

```diff
@@ -17,118 +17,273 @@ open FSharp.Editor.Tests.Helpers
         let sourceText = SourceText.From code
+        let! fsharpDiagnostics = FSharpDiagnostics.generate mode document
 
-        let! context = createTestCodeFixContext code mode fixProvider
+        let canFix =
+            match mode with
+            | Manual _ -> fun _ -> true
+            | _ -> fixProvider.CanFix
 
-        let! result = fixProvider.GetCodeFixIfAppliesAsync context
+        let context =
+            fsharpDiagnostics
+            |> Seq.map (Diagnostic.ofFSharpDiagnostic sourceText document.FilePath)
+            |> Seq.toImmutableArray
+            |> CodeFixContext.tryCreate canFix document
 
-        return
-            (result
-             |> toOption
-             |> Option.map (fun codeFix ->
-                 {
-                     Message = codeFix.Message
-                     FixedCode = (sourceText.WithChanges codeFix.Changes).ToString()
-                 }))
+        return!
+            context
+            |> ValueOption.either fixProvider.GetCodeFixIfAppliesAsync (CancellableTask.singleton ValueNone)
+            |> CancellableTask.map (ValueOption.map (TestCodeFix.ofFSharpCodeFix sourceText) >> ValueOption.toOption)
     }
-    |> CancellableTask.startWithoutCancellation
-    |> fun task -> task.Result
+    |> CancellableTask.runSynchronouslyWithoutCancellation
 
-let multiFix (code: string) mode (fixProvider: 'T :> IFSharpMultiCodeFixProvider) =
+let multiFix (code: string) mode (fixProvider: 'T when 'T :> IFSharpMultiCodeFixProvider and 'T :> CodeFixProvider) =
     cancellableTask {
+        let document = Document.create mode code
         let sourceText = SourceText.From code
+        let! fsharpDiagnostics = FSharpDiagnostics.generate mode document
 
-        let! context = createTestCodeFixContext code mode fixProvider
+        let canFix =
+            match mode with
+            | Manual _ -> fun _ -> true
+            | _ -> fixProvider.CanFix
 
-        let! result = fixProvider.GetCodeFixesAsync context
+        let context =
+            fsharpDiagnostics
+            |> Seq.map (Diagnostic.ofFSharpDiagnostic sourceText document.FilePath)
+            |> Seq.toImmutableArray
+            |> CodeFixContext.tryCreate canFix document
 
-        return
-            result
-            |> Seq.map (fun codeFix ->
-                {
-                    Message = codeFix.Message
-                    FixedCode = (sourceText.WithChanges codeFix.Changes).ToString()
-                })
+        return!
+            context
+            |> ValueOption.either fixProvider.GetCodeFixesAsync (CancellableTask.singleton Seq.empty)
+            |> CancellableTask.map (Seq.map (TestCodeFix.ofFSharpCodeFix sourceText))
     }
-    |> CancellableTask.startWithoutCancellation
-    |> fun task -> task.Result
+    |> CancellableTask.runSynchronouslyWithoutCancellation
+
+/// Contains types and functions for coveniently making code fix assertions using xUnit.
+[<AutoOpen>]
```

Comment:
This should probably go to the `FSharp.Test.Utilities` project, the module `Assert.fs` is there basically for these purposes.

Reply:
Hmm. They could, although they're somewhat tailored to code fix tests specifically (the `MissingCodeFixException` versus `UnexpectedCodeFixException` versus `WrongCodeFixException`, etc.).

The `memberData` CE _could_ be more broadly useful, although other uses might want additional overloads (taking triples instead of pairs), etc.

Reply:
Sure, we can move this once there is actually a need for reusing this code.